cdc/processor/pipeline/cyclic_mark.go

```diff
@@ -77,87 +79,81 @@ func (n *cyclicMarkNode) Receive(ctx pipeline.NodeContext) error {
 	}
-	err := event.WaitPrepare(ctx)
-	if err != nil {
-		return errors.Trace(err)
-	}
 	markRow := event.Row
 	if markRow == nil {
-		return nil
+		return
 	}
 	replicaID := extractReplicaID(markRow)
 	// Establishing the mapping from StartTs to ReplicaID
 	n.currentReplicaIDs[markRow.StartTs] = replicaID
-	if rows, exist := n.rowsUnknownReplicaID[markRow.StartTs]; exist {
-		// the replicaID of these rows we did not know before, but now we know through received mark row now.
-		delete(n.rowsUnknownReplicaID, markRow.StartTs)
-		n.sendNormalRowToNextNode(ctx, replicaID, rows...)
+	if events, exist := n.unknownReplicaIDEvents[markRow.StartTs]; exist {
+		// the replicaID of these events we did not know before, but now we know through received mark row now.
+		delete(n.unknownReplicaIDEvents, markRow.StartTs)
+		n.sendNormalRowEventToNextNode(ctx, replicaID, events...)
 	}
-	return nil
 }
 
 func (n *cyclicMarkNode) flush(ctx pipeline.NodeContext, commitTs uint64) {
 	if n.currentCommitTs == commitTs {
 		return
 	}
-	// all mark rows and normal rows in current transaction is received now.
-	// there are still unmatched normal rows in the cache, their replicaID should be local replicaID.
-	for _, rows := range n.rowsUnknownReplicaID {
-		for _, row := range rows {
+	// all mark events and normal events in current transaction is received now.
+	// there are still unmatched normal events in the cache, their replicaID should be local replicaID.
+	for _, events := range n.unknownReplicaIDEvents {
+		for _, row := range events {
 			row.ReplicaID = n.localReplicaID
 		}
-		n.sendNormalRowToNextNode(ctx, n.localReplicaID, rows...)
+		n.sendNormalRowEventToNextNode(ctx, n.localReplicaID, events...)
 	}
-	if len(n.rowsUnknownReplicaID) != 0 {
-		n.rowsUnknownReplicaID = make(map[model.Ts][]*model.PolymorphicEvent)
+	if len(n.unknownReplicaIDEvents) != 0 {
+		n.unknownReplicaIDEvents = make(map[model.Ts][]*model.PolymorphicEvent)
 	}
 	if len(n.currentReplicaIDs) != 0 {
 		n.currentReplicaIDs = make(map[model.Ts]uint64)
 	}
 	n.currentCommitTs = commitTs
 }
 
-// sendNormalRowToNextNode filter the specified normal rows by the FilterReplicaID config item, and send rows to the next node.
-func (n *cyclicMarkNode) sendNormalRowToNextNode(ctx pipeline.NodeContext, replicaID uint64, rows ...*model.PolymorphicEvent) {
+// sendNormalRowEventToNextNode filter the specified normal row events by the FilterReplicaID config item,
+// and send events to the next node.
+func (n *cyclicMarkNode) sendNormalRowEventToNextNode(ctx pipeline.NodeContext, replicaID uint64, events ...*model.PolymorphicEvent) {
 	if _, shouldFilter := n.filterReplicaID[replicaID]; shouldFilter {
 		return
 	}
-	for _, row := range rows {
-		row.ReplicaID = replicaID
-		ctx.SendToNextNode(pipeline.PolymorphicEventMessage(row))
+	for _, event := range events {
+		event.ReplicaID = replicaID
```

Comment:
Is `event.ReplicaID` useful after this PR

Reply:
Good catch! Let me try to remove it.

Reply:
Done.